celo-org/celo-monorepo
# Twilio Verify API

### Description

The Twilio Verify Service is a managed service for sending N-digit codes globally. Because there's no risk of sending marketing material, Twilio can automatically manage regulations for each recipient geography on behalf of the sender. The service also manages localization for select markets. We chose Twilio as the first SMS provider to measure the impact of switching to a Verify Service since it 1) is heavily used by many operators 2) supports sending custom codes which minimizes Valora-side changes.

Changes to note:
- Changes SMS text content
  - Android  
    - from `<#> Celo security code is: 12345678 bU9E4ctGtIW` 
    - to `<#>Your Celo verification code is: 12345678 bU9E4ctGtIW`
  - iOS
    - from `Celo security code is: 12345678` 
    - to `Your Celo verification code is: 12345678`
- Expands localized markets from 3 to 36
- Doesn't support a call-back URL which will affect some metrics on our dashboard
- Supporting verification for landlines (using voice)

Pre-requisite steps for Attestation Service operators:
- Create a new Verify Service in the portal
- Create a support ticket to enable custom security codes

### Other changes

We now include `language` and `appSignature` in the `AttestationModel` object. The Verify API uses these directly to adjust the messaging. Added a new migration file for adding these columns to the Attestations table.

Added `twilioVerifySidProvided` to the /status endpoint and tracking via metadata-crawler.

### Tested

Tested manually in Alfajores using Valora Alfajores app:
 - Android device receives app hash and can auto-read message
 - Can re-request SMS from same services
 - Successfully retrieves signature from AS using security code

### Related issues

- Fixes #8145 

### Backwards compatibility

`AttestationModel` objects that are previously stored in the DB do not have the new properties (`language` and `appSignature). When reading the `AttestationModel`, these new properties are be added for compatibility.

### Documentation

Updated Twilio guide
